### ISSUE TITLE
Undefined behaviour snprintf compiler warnings (errors) with GCC 12

### ISSUE THREAD

**Issue by tokyovigilante** [swaywm/sway#6784]
### Please read the following before submitting:
- Please do NOT submit bug reports for questions. Ask questions on IRC at #sway on Libera Chat.
- Proprietary graphics drivers, including nvidia, are not supported. Please use the open source equivalents, such as nouveau, if you would like to use Sway.
- Please do NOT submit issues for information from the github wiki. The github wiki is community maintained and therefore may contain outdated information, scripts that don't work or obsolete workarounds.
  If you fix a script or find outdated information, don't hesitate to adjust the wiki page.

### Please fill out the following:
- **Sway Version:**
sway HEAD@cd1ee0e/wlroots HEAD

- **Description:**
I upgraded to Fedora rawhide due to Sway 1.7+'s requirement for wayland >-1.20, which includes GCC 12. This raises the following build warnings:

```
ninja: Entering directory `build'
[493/539] Compiling C object sway/sway.p/commands_output_background.c.o
../sway/commands/output/background.c: In function ‘output_cmd_background’:
../sway/commands/output/background.c:115:25: warning: ‘sprintf’ argument 4 overlaps destination object ‘src’ [-Wrestrict]
  115 |                         sprintf(src, "%s/%s", conf_path, rel_path);
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../sway/commands/output/background.c:81:23: note: destination object referenced by ‘restrict’-qualified argument 1 was declared here
   81 |                 char *src = join_args(argv, j);
      |                       ^~~
[535/539] Compiling C object sway/sway.p/server.c.o
../sway/server.c: In function ‘server_init’:
../sway/server.c:217:50: warning: ‘%d’ directive writing between 1 and 11 bytes into a region of size 8 [-Wformat-overflow=]
  217 |                 sprintf(name_candidate, "wayland-%d", i);
      |                                                  ^~
../sway/server.c:217:41: note: directive argument in the range [-2147483647, 32]
  217 |                 sprintf(name_candidate, "wayland-%d", i);
      |                                         ^~~~~~~~~~~~
../sway/server.c:217:17: note: ‘sprintf’ output between 10 and 20 bytes into a destination of size 16
  217 |                 sprintf(name_candidate, "wayland-%d", i);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[539/539] Linking target sway/sway
```
`gcc -v`:
```
Using built-in specs.
COLLECT_GCC=/usr/bin/gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/12/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,objc,obj-c++,ada,go,d,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --with-isl=/builddir/build/BUILD/gcc-12.0.1-20220118/obj-x86_64-redhat-linux/isl-install --enable-offload-targets=nvptx-none --without-cuda-driver --enable-offload-defaulted --enable-gnu-indirect-function --enable-cet --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux --with-build-config=bootstrap-lto --enable-link-serialization=1
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 12.0.1 20220118 (Red Hat 12.0.1-0) (GCC)
```

**Comment by r-c-f** [comment]
I would call these fairly egregious false positives. In the first case `src` is assigned a new value before it would alias, and in the second case gcc is ignoring the initial assignment and assuming `INT_MIN` despite guessing the upper bound correctly (32)

**Comment by bl4ckb0ne** [comment]
It would not hurt to move those calls to `snprintf`. Patches welcome.

**Comment by alebastr** [comment]
`snprintf` might just change the warning to `Wstringop-truncation` or `Wformat-truncation`. Using a smaller type for the variable would be a better change in the second case.

GCC 12 is an early development snapshot though and any warnings or diagnostics are not stable yet. I'd wait until at least GCC feature freeze before trying to fix anything.

**Comment by emersion** [comment]
Fixed in #6802

**Comment by tokyovigilante** [comment]
Thanks!
